### PR TITLE
Make train.py a driver on a loadable module Trainer.py

### PR DESCRIPTION
**Pull request by JianyuZhan**
This PR is a step forward to achieve the goal in https://github.com/OpenNMT/OpenNMT-py/issues/239.

Now the train code are a loadable module in `onmt/Trainer.py`. User can load it and use it as pytorch-style, just as `train.py` does(`train.py` itself now acts like a user of this module too):

```
def train_model(model, train_data, valid_data, fields, optim):

    trainer = onmt.Trainer(model, train_data, valid_data, fields, optim,
                           opt.batch_size, opt.gpuid, opt.copy_attn,
                           opt.copy_attn_force, opt.truncated_decoder,
                           opt.max_generator_batches)

    for epoch in range(opt.start_epoch, opt.epochs + 1):
        print('')

        # 1. Train for one epoch on the training set.
        train_stats = trainer.train(epoch, report_func)
        print('Train perplexity: %g' % train_stats.ppl())
        print('Train accuracy: %g' % train_stats.accuracy())

        # 2. Validate on the validation set.
        valid_stats = trainer.validate()
        print('Validation perplexity: %g' % valid_stats.ppl())
        print('Validation accuracy: %g' % valid_stats.accuracy())

        # 3. Log to remote server.
        if opt.exp_host:
            train_stats.log("train", experiment, optim.lr)
            valid_stats.log("valid", experiment, optim.lr)

        # 4. Update the learning rate
        trainer.epoch_step(valid_stats.ppl(), epoch)

        # 5. Drop a checkpoint if needed.
        if epoch >= opt.start_checkpoint_at:
            trainer.drop_checkpoint(opt, epoch, valid_stats)
```

The current constructor interface is still not minimal, we can move out the `loss criterion` staff. Since now other people are working on loss refactor work.So I plan to leave it as is. Anyway, the interface is quite neat and clear now.

I also refactor the `main()` method in `train.py`, it now is more tidy and easy to understand too. See https://github.com/JianyuZhan/OpenNMT-py/blob/train_refacotr/train.py#L191

**Patch list**
Total 10 patches in this PR.

First two are cleanup patches.
1. 18eeaa7 train.py: move the leftover opt definition to opts.py
2. c71256b train.py: use collect_features api

This one refactors  `main()` method.
3. a3a70c8 train.py: factor out the skeleton of main function

This one forms the skeleton of the `onmt/Trainer.py`, with two interface `train()` and `validate()`, following the pytorch style.
4. 10822c7 Trainer: add onmt/Trainer.py to abstract out the main training logic

These two add two more interfaces:  `epoch_step()`, called every epoch for update lr(I didn't call it `step()` as pytorch does because we have a batch level `step()` too).  `drop_checkpoint()`, called conditionally to save a snapshot.
5. 5844c49 Trainer: add an epoch_step method for update lr every epoch
6. 6d10740 Trainer: add a drop_checkpoint method

Based on previous one, this simplifies the `train.py`, making it the first user of `Trainer.py`
7. 9f8f1db train.py: abstract the train_model logic with Trainer methods.

These two move `Statistics` class to its own file `onmt/Statistics.py`
8. a5d5de3 Statistics: move Statistics to its own file onmt/Statistics
9. 585ea93 Statistics: shrink log method interface to accept lr instead of optim

This shrinks the `Trainer` constructor a bit, the `Crayon logging` should be part of `train.py`, not `Trainer.py`, so I pass in a `report_func()` function to `trainer.train()` to do this duty, instead of letting `Trainer.py` to handle it itself. Now `trainer.py` is unware of `Crayon logging` thingy.
10. 8c39e8c Trainer: shrink constructor

**Test done**
1. `test/pull_request_chk.sh` passed.
2. A standard RNN `preprocess + train + translate` test passed.
3. `transformer`  `preprocess + train + translate` test passed.
4. `transformer train_from` test passed.

### CODE REVIEW

File: onmt/Trainer.py [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+        self.fields = fields
+        self.optim = optim
+        self.batch_size = batch_size
+        self.gpuid = gpuid
+        self.copy_attn = copy_attn
+        self.truncated_decoder = truncated_decoder
+        self.max_generator_batches = max_generator_batches
+
+        # Define criterion.
+        padding_idx = fields['tgt'].vocab.stoi[onmt.IO.PAD_WORD]
```
Comment (srush) [inline onmt/Trainer.py]: Remove all this copy stuff from this class. Ideally the trainer should not even know about NMT details at all.

File: onmt/Trainer.py [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+            dataset=valid_data, batch_size=batch_size,
+            device=gpuid[0] if gpuid else -1,
+            train=False, sort=True)
+
+        # Set model in training mode.
+        self.model.train()
+
+    def train(self, epoch, report_func=None):
+        """ Called for each epoch to train. """
+        closs = onmt.Loss.LossCompute(self.model.generator, self.criterion,
```
Comment (srush) [inline onmt/Trainer.py]: This part should be passed in somehow.

File: onmt/Statistics.py [hunk 3]
```diff
@@ -0,0 +1,51 @@
+"""
+Statistics is a class used to store statistics data during training.
+"""
+import time
+import sys
+import math
+
+
+class Statistics(object):
```
Comment (srush) [inline onmt/Statistics.py]: I am okay with this going in Trainer.py if you want.

File: train.py [hunk 4]
```diff
@@ -61,158 +57,58 @@
...
-        src = onmt.IO.make_features(batch, 'src')
-        tgt = onmt.IO.make_features(batch, 'tgt')
-        outputs, attn, _ = model(src, tgt, src_lengths)
-        gen_state = loss.make_loss_batch(
-            outputs, batch, attn, (0, batch.tgt.size(0)))
-        _, batch_stats = loss.compute_loss(batch=batch, **gen_state)
-        stats.update(batch_stats)
-    model.train()
-    return stats
+def report_func(epoch, batch, num_batches,
```
Comment (srush) [inline train.py]: I like this!

File: onmt/Trainer.py [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+        # Basic attributes.
+        self.model = model
+        self.train_data = train_data
+        self.valid_data = valid_data
+        self.fields = fields
+        self.optim = optim
+        self.batch_size = batch_size
+        self.gpuid = gpuid
+        self.copy_attn = copy_attn
+        self.truncated_decoder = truncated_decoder
```
Comment (srush) [inline onmt/Trainer.py]: Oh sorry, this is a bad name. It is an integer, so name should indicate it is a length.

File: onmt/Trainer.py [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+        # Define criterion.
+        padding_idx = fields['tgt'].vocab.stoi[onmt.IO.PAD_WORD]
+        if not copy_attn:
+            self.criterion = onmt.Loss.nmt_criterion(
+                len(fields['tgt'].vocab), gpuid, padding_idx)
+        else:
+            self.criterion = onmt.modules.CopyCriterion(
+                len(fields['tgt'].vocab), copy_attn_force, padding_idx)
+
+        # Create a train data iterator.
```
Comment (srush) [inline onmt/Trainer.py]: Actually, I think these should be created on the fly each epoch. Train and valid should take train_data and valid_data as arguments. Let's users implement things like curriculum

**Comment by JianyuZhan** [comment]
v1->v2:

- created `NMTLossCompute` and `CopyGeneratorLossCompute` classes(inherited from `LossComputeBase`), and hide the criterion inside,  also leverage `nn.Module` to use `cuda` behavior.
- moved out the loss compute stuff,now `Trainer` accepts one of these two class as `train_loss`/`valid_loss` for computing loss. 
- moved out the iterator stuff, now `Trainer` accepts `train_iter` and `valid_iter` for iterating over data.
- simplified the `Trainer` constructor to minimal.
- tidy  up the `Trainer.py` code path, now the shard compute loss code is replaced with a `sharded_compute_loss()`.
- fixed a `batch has no alignment attribute` bug due to `-dynamic_dict` is not passed for the `-copy_attn` case.

This refactor work aims at totally separate out mechanism and strategy from `Tranier.py`. Now its interface is like:

```
def train_model(model, train_data, valid_data, fields, optim):

    train_iter = make_train_data_iter(train_data, opt)
    valid_iter = make_valid_data_iter(valid_data, opt)

    train_loss = make_loss_compute(model, fields["tgt"].vocab,
                                   train_data, opt)
    valid_loss = make_loss_compute(model, fields["tgt"].vocab,
                                   valid_data, opt)

    trunc_size = opt.truncated_decoder  # Badly named...
    shard_size = opt.max_generator_batches

    trainer = onmt.Trainer(model, train_iter, valid_iter,
                           train_loss, valid_loss, optim,
                           trunc_size, shard_size)

    for epoch in range(opt.start_epoch, opt.epochs + 1):
        print('')

        # 1. Train for one epoch on the training set.
        train_stats = trainer.train(epoch, report_func)
        print('Train perplexity: %g' % train_stats.ppl())
        print('Train accuracy: %g' % train_stats.accuracy())

        # 2. Validate on the validation set.
        valid_stats = trainer.validate()
        print('Validation perplexity: %g' % valid_stats.ppl())
        print('Validation accuracy: %g' % valid_stats.accuracy())

        # 3. Log to remote server.
        if opt.exp_host:
            train_stats.log("train", experiment, optim.lr)
            valid_stats.log("valid", experiment, optim.lr)

        # 4. Update the learning rate
        trainer.epoch_step(valid_stats.ppl(), epoch)

        # 5. Drop a checkpoint if needed.
        if epoch >= opt.start_checkpoint_at:
            trainer.drop_checkpoint(opt, epoch, fields, valid_stats)
```

**Test done**

1. Standard RNN preprocess + train + translate test passed.
2. transformer preprocess + train + translate test passed, with `train_from` tested passed.
3.   **`copy_attn` case not done yet,** due to a current existing bug.
